src/control/lib/hardware/topology.go

```diff
@@ -34,20 +41,130 @@ func (t *Topology) AllDevices() map[string]*Device {
 	return devsByName
 }
 
+func PrintTopology(t *Topology, output io.Writer) error {
+	if t == nil {
+		fmt.Fprintf(output, "No topology information available\n")
+	}
+
+	for _, numaNode := range t.NUMANodes {
+		coreSet := &system.RankSet{}
+		for _, core := range numaNode.Cores {
+			coreSet.Add(system.Rank(core.ID))
+		}
+
+		fmt.Fprintf(output, "NUMA node %d (%s)\n", numaNode.ID, numaNode.Set)
+		fmt.Fprintf(output, "  CPU cores: %s\n", coreSet)
+		fmt.Fprintf(output, "  PCI devices: \n")
+		for addr, devs := range numaNode.PCIDevices {
+			fmt.Fprintf(output, "    %s\n", &addr)
+			for _, dev := range devs {
+				fmt.Fprintf(output, "      %s\n", dev)
+			}
+		}
+		fmt.Fprintf(output, "  PCI buses: \n")
+		for _, bus := range numaNode.PCIBuses {
+			fmt.Fprintf(output, "    %s\n", bus)
+		}
+	}
+
+	return nil
+}
+
 type (
+	CPUCore struct {
+		ID       uint      `json:"id"`
+		NUMANode *NUMANode `json:"-"`
+	}
+
+	// PCIDevice represents an individual hardware device.
+	PCIDevice struct {
+		Name      string            `json:"name"`
+		Type      DeviceType        `json:"type"`
+		NUMANode  *NUMANode         `json:"-"`
+		PCIAddr   common.PCIAddress `json:"pci_address"`
+		LinkSpeed float64           `json:"link_speed"`
```

Comment:
Would be a good idea to put the units in the name, so it's clear.

Reply:
Hmm. It's not in the name on the C side, there's just a comment in the header that the value represents GB/s. I was going to say that if it changes later we'll have to add some compatibility logic, but I guess we'd have to do that anyhow...

Reply:
Yeah. I'm just thinking for clarity and stability for anyone who has to use this function. As-is I had to look at the format it's printed in to figure out the units.

Reply:
we should have at least a comment if not IMO